### PR TITLE
[LETS-143] Add the type of server to the client-server connections

### PR DESCRIPTION
**Pull request by adrianbrab**
http://jira.cubrid.org/browse/LETS-143

Client need to specify the desired server typoe when making a connection. Cub_master already differentiates the servers between them by type, the change is to make the distinction when returning the actual connection.

All clients should be able to connect to the transaction server. Some client types should be able to connected to the page server too. These clients need to pass along to cub_master the desired type of server.

The clients that can connect to the page server too: backupdb, paramdump, statdump, csql with --sysadm.

### CODE REVIEW

File: src/communication/communication_server_channel.cpp [hunk 1]
```diff
@@ -77,8 +79,15 @@ namespace cubcomm
 		  " server_name_size:%zu\n",
 		  cmd_type, hostname, port, m_server_name.c_str (), m_server_name.size ());
 
+    /*
+     * The packing of the server name and type is done by setting the first
+     * character to the numer related to the server type enum value, the rest of the
+     * buffer space is used to copy the server name.
+     */
+    msg = (char) (m_server_type + '0');
```
Comment (ribram) [inline src/communication/communication_server_channel.cpp]:
```suggestion
    msg = ((char) m_server_type) + '0';
```

File: src/connection/connection_cl.c [hunk 2]
```diff
@@ -1027,8 +1028,15 @@ css_connect_to_cubrid_server (char *host_name, char *server_name)
 
   timeout = prm_get_integer_value (PRM_ID_TCP_CONNECTION_TIMEOUT) * 1000;
 
+  /*
+   * The packing of the server name and type is done by setting the first
+   * character to the numer related to the server type enum value, the rest of the
+   * buffer space is used to copy the server name.
+   */
+  msg = (char) (server_type + '0');
```
Comment (ribram) [inline src/connection/connection_cl.c]:
```suggestion
  msg = ((char) server_type) + '0';
```

File: src/connection/connection_cl.c [hunk 3]
```diff
@@ -1027,8 +1028,15 @@ css_connect_to_cubrid_server (char *host_name, char *server_name)
...
+  msg.append (server_name, strlen (server_name));
```
Comment (ribram) [inline src/connection/connection_cl.c]:
```suggestion
  msg.append (server_name, std::strlen (server_name));
```

File: src/executables/master.c [hunk 4]
```diff
@@ -652,13 +652,14 @@ static void
...
 {
   SOCKET_QUEUE_ENTRY *temp;
-  char *server_name = NULL;
   int name_length, buffer;
   name_length = 1024;
-  if (css_receive_data (conn, rid, &server_name, &name_length, -1) == NO_ERRORS && server_name != NULL)
+  std::string server_name;
+  SERVER_TYPE type;
+
+  if (receive_server_info (conn, rid, server_name, type) == NO_ERRORS && !server_name.empty ())
```
Comment (ribram) [inline src/executables/master.c]: It is best to separate functions that modify one of the args and then using the same arg again in different conditionals. It is easier to read.
```suggestion

  if (receive_server_info (conn, rid, server_name, type) != NO_ERRORS)
    {
      css_free_conn (conn);
      return;
    }

  if (!server_name.empty ())
```

File: src/executables/master.c [hunk 5]
```diff
@@ -681,20 +681,19 @@ css_send_to_existing_server (CSS_CONN_ENTRY * conn, unsigned short rid, CSS_SERV
...
 		    {
-		      free_and_init (server_name);
 		      css_free_conn (conn);
 		      return;
 		    }
 		  else if (!temp->ha_mode)
 		    {
-		      temp = css_return_entry_of_server (server_name, css_Master_socket_anchor, SERVER_TYPE_ANY);
+		      temp =
+			css_return_entry_of_server (server_name.c_str (), css_Master_socket_anchor, SERVER_TYPE_ANY);
```
Comment (ribram) [inline src/executables/master.c]: SERVER_TYPE_ANY or type?

File: src/executables/utility.h [hunk 6]
```diff
@@ -1059,6 +1059,8 @@ typedef struct _ha_config
 #define BACKUP_SLEEP_MSECS_L                    "sleep-msecs"
 #define BACKUP_SEPARATE_KEYS_S                  'k'
 #define BACKUP_SEPARATE_KEYS_L                  "separate-keys"
+#define BACKUP_PAGE_SERVER_S                    34555
```
Comment (ribram) [inline src/executables/utility.h]: Increment from highest backup option; in this case, who added BACKUP_SLEEP_MSECS_S made a mistake. Modify that to 10507
```suggestion
#define BACKUP_PAGE_SERVER_S                    10508
```

File: src/executables/utility.h [hunk 7]
```diff
@@ -1387,6 +1389,8 @@ typedef struct _ha_config
 #define CSQL_QUERY_COLUMN_ENCLOSURE_L		"enclosure"
 #define CSQL_LOADDB_OUTPUT_S			'd'
 #define CSQL_LOADDB_OUTPUT_L			"loaddb-output"
+#define CSQL_PAGE_SERVER_S                      34556
```
Comment (ribram) [inline src/executables/utility.h]:
```suggestion
#define CSQL_PAGE_SERVER_S                      10219
```

File: src/executables/utility.h [hunk 8]
```diff
@@ -1444,6 +1448,8 @@ typedef struct _ha_config
 #define PARAMDUMP_SA_MODE_L                     "SA-mode"
 #define PARAMDUMP_CS_MODE_S                     'C'
 #define PARAMDUMP_CS_MODE_L                     "CS-mode"
+#define PARAMDUMP_PAGE_SERVER_S                 34557
```
Comment (ribram) [inline src/executables/utility.h]:
```suggestion
#define PARAMDUMP_PAGE_SERVER_S                 12200
```

File: src/executables/utility.h [hunk 9]
```diff
@@ -1454,6 +1460,8 @@ typedef struct _ha_config
 #define STATDUMP_CUMULATIVE_L                   "cumulative"
 #define STATDUMP_SUBSTR_S			's'
 #define STATDUMP_SUBSTR_L			"substr"
+#define STATDUMP_PAGE_SERVER_S                  34558
```
Comment (ribram) [inline src/executables/utility.h]:
```suggestion
#define STATDUMP_PAGE_SERVER_S                  12300
```

File: src/transaction/boot_cl.h [hunk 10]
```diff
@@ -76,4 +77,6 @@ extern void boot_clear_host_connected (void);
 extern char *boot_get_server_session_key (void);
 extern void boot_set_server_session_key (const char *key);
 extern int boot_add_collations (MOP class_mop);
+extern SERVER_TYPE get_conn_server_type ();
+extern void set_conn_server_type (SERVER_TYPE server_type);
```
Comment (ribram) [inline src/transaction/boot_cl.h]: Rename using the boot_ suffix specific to the boot module:
```suggestion
extern SERVER_TYPE boot_get_conn_server_type ();
extern void boot_set_conn_server_type (SERVER_TYPE server_type);
```

**Review by ribram — COMMENTED** [review]
Rename the server_name argument as message in css_server_connect and css_common_connect.

File: src/transaction/boot_cl.h (hunk 10 above)
Comment (adrianbrab) [reply src/transaction/boot_cl.h]: renamed

File: src/executables/utility.h (hunk 7 above)
Comment (adrianbrab) [reply src/executables/utility.h]: did you mean 12020, as the previous one `CSQL_QUERY_COLUMN_ENCLOSURE_S` has the value 12019

File: src/executables/master.c (hunk 5 above)
Comment (adrianbrab) [reply src/executables/master.c]: should it not be the type delivered by the client via the connection?

Comment (adrianbrab) [reply src/executables/master.c]: so type

File: src/executables/master.c (hunk 4 above)
Comment (adrianbrab) [reply src/executables/master.c]: also changed to `server_name.empty () == false` for better visibility

File: src/connection/connection_cl.c [hunk 11]
```diff
@@ -104,9 +105,9 @@ static void css_dealloc_conn (CSS_CONN_ENTRY * conn);
 
 static int css_read_header (CSS_CONN_ENTRY * conn, NET_HEADER * local_header);
 static CSS_CONN_ENTRY *css_common_connect (const char *host_name, CSS_CONN_ENTRY * conn, int connect_type,
-					   const char *server_name, int server_name_length, int port, int timeout,
+					   const char *message, int server_name_length, int port, int timeout,
```
Comment (ribram) [inline src/connection/connection_cl.c]:
```suggestion
					   const char *message, int message_length, int port, int timeout,
```

File: src/connection/connection_cl.c [hunk 12]
```diff
@@ -717,7 +718,7 @@ css_receive_error (CSS_CONN_ENTRY * conn, unsigned short req_id, char **buffer,
  *   rid(out):
  */
 static CSS_CONN_ENTRY *
-css_common_connect (const char *host_name, CSS_CONN_ENTRY * conn, int connect_type, const char *server_name,
+css_common_connect (const char *host_name, CSS_CONN_ENTRY * conn, int connect_type, const char *message,
 		    int server_name_length, int port, int timeout, unsigned short *rid, bool send_magic)
```
Comment (ribram) [inline src/connection/connection_cl.c]:
```suggestion
		    int message_length, int port, int timeout, unsigned short *rid, bool send_magic)
```

File: src/connection/connection_cl.c [hunk 13]
```diff
@@ -745,7 +746,7 @@ css_common_connect (const char *host_name, CSS_CONN_ENTRY * conn, int connect_ty
 	  return NULL;
 	}
 
-      if (css_send_request (conn, connect_type, rid, server_name, server_name_length) == NO_ERRORS)
+      if (css_send_request (conn, connect_type, rid, message, server_name_length) == NO_ERRORS)
```
Comment (ribram) [inline src/connection/connection_cl.c]:
```suggestion
      if (css_send_request (conn, connect_type, rid, message, message_length) == NO_ERRORS)
```

File: src/connection/connection_cl.c [hunk 14]
```diff
@@ -1027,8 +1029,15 @@ css_connect_to_cubrid_server (char *host_name, char *server_name)
 
   timeout = prm_get_integer_value (PRM_ID_TCP_CONNECTION_TIMEOUT) * 1000;
 
+  /*
+   * The packing of the server name and type is done by setting the first
+   * character to the numer related to the server type enum value, the rest of the
+   * buffer space is used to copy the server name.
+   */
+  msg = ((char) server_type) + '0';
+  msg.append (server_name, std::strlen (server_name));
```
Comment (ribram) [inline src/connection/connection_cl.c]: css_server_connect calls `strlen` again on `msg.c_str()`, therefore msg must end with a null-terminated character. Appending `strlen` will miss the ending null-terminated character. Either append `strlen+1` or append the `server_name` without any length mentioned to automatically copy until the null-terminated character.

**Review by ribram — APPROVED** [review]

File: src/communication/communication_server_channel.cpp [hunk 15]
```diff
@@ -29,10 +29,11 @@
 namespace cubcomm
 {
 
-  server_channel::server_channel (const char *server_name, int max_timeout_in_ms)
+  server_channel::server_channel (const char *server_name, SERVER_TYPE server_type, int max_timeout_in_ms)
     : channel (max_timeout_in_ms),
       m_server_name (server_name)
   {
+    m_server_type = server_type;
```
Comment (cristiarg) [inline src/communication/communication_server_channel.cpp]: move to init list

File: src/communication/communication_server_channel.cpp [hunk 16]
```diff
@@ -29,10 +29,11 @@
...
     assert (server_name != NULL);
```
Comment (cristiarg) [inline src/communication/communication_server_channel.cpp]: new code in hpp/cpp should use `nullptr` all over; macro `NULL` remains for h/c files

File: src/communication/communication_server_channel.cpp [hunk 17]
```diff
@@ -54,6 +55,7 @@ namespace cubcomm
   {
     unsigned short m_request_id;
     css_error_code rc = NO_ERRORS;
+    std::string msg;
```
Comment (cristiarg) [inline src/communication/communication_server_channel.cpp]: In new - c++ - code, I would declare all variables closest to the scope/place of their usage.

File: src/executables/master.c [hunk 18]
```diff
@@ -652,13 +652,20 @@ static void
 css_send_to_existing_server (CSS_CONN_ENTRY * conn, unsigned short rid, CSS_SERVER_REQUEST request)
 {
   SOCKET_QUEUE_ENTRY *temp;
-  char *server_name = NULL;
   int name_length, buffer;
   name_length = 1024;
-  if (css_receive_data (conn, rid, &server_name, &name_length, -1) == NO_ERRORS && server_name != NULL)
+  std::string server_name;
+  SERVER_TYPE type;
```
Comment (cristiarg) [inline src/executables/master.c:658]: I would not let any variable uninitialized.

**Review by cristiarg — APPROVED** [review]
Pedantic remarks. Fell free to ignore them.

**Pull request merged**